compiler/lquantizer/src/LQBinarizer.cpp

```diff
@@ -0,0 +1,192 @@
+  quantization_thresholds = std::make_unique<float[]>(quantized_levels);
+
+  // init quantization levels
+  {
+    for (int32_t i = 0; i < quantized_levels; ++i)
+    {
+      // TODO investigate faster variants to perform this
+      quantization_thresholds[i] = 0.0f;
+      for (int32_t b = 0; b < encode_bits; ++b)
+      {
+        quantization_thresholds[i] += (i & (1 << b)) ? scales[b] : -scales[b];
+      }
+    }
+  }
+
+  // compute quantization thresholds via levels
+  {
+    float last_v = quantization_thresholds[0];
+    quantization_thresholds[0] = std::numeric_limits<float>::lowest(); // -INF
+
+    for (uint32_t i = 1; i < quantized_levels; ++i)
+    {
+      float curr_v = quantization_thresholds[i];
+      quantization_thresholds[i] = (curr_v + last_v) / 2.0f;
+      last_v = curr_v;
+    }
+  }
+}
+
+// this bin search returns t[l] for values in (t[l], t[l+1]]
+int32_t LQBinarizer::bin_search_encoding(float value)
+{
+  int32_t left = 0, right = quantized_levels - 1;
+
+  while (left < right)
+  {
+    uint32_t middle = (left + right + 1) >> 1;
+
+    if (quantization_thresholds[middle] < value)
+    {
+      left = middle;
+    }
+    else
+    {
+      right = middle - 1;
+    }
+  }
+
+  return left;
+}
+
+void LQBinarizer::quantize_and_pack(const float *data_vector)
+{
+  Shape shape = {encode_bits, data_bin_size};
+  memset(data_binary.get(), 0, shape.num_elements() * sizeof(int32_t));
+
+  for (int32_t i = 0; i < data_float_size; ++i)
+  {
+    // encode value of input
+    int32_t encoding = bin_search_encoding(data_vector[i]);
+
+    // pack value to binary buffer
+    {
+      int32_t idx = i >> 5;    // divide to 32
+      int32_t offset = i & 31; // mod of 32
+      for (int32_t b = 0; b < encode_bits; ++b)
+      {
+        data_binary[calcOffset(shape, b, idx)] |= ((encoding >> b) & 1) << offset;
+      }
+    }
+  }
+}
+
+std::vector<float> LQBinarizer::dequantize()
+{
+  std::vector<float> dequantized_data(data_float_size, 0);
+
+  // calculate data from binary
+  for (int32_t i = 0; i < data_float_size; ++i)
+  {
+    int32_t idx = i >> 5;    // divide to 32
+    int32_t offset = i & 31; // mod of 32
+
+    for (int32_t b = 0; b < encode_bits; ++b)
+    {
+      int32_t bit_value = (data_binary[b * data_bin_size + idx] >> offset) & 1;
+      dequantized_data[i] += (bit_value == 1) ? scales[b] : -scales[b];
+    }
+  }
+
+  return dequantized_data;
+}
+
+// Method returns bits in float representation for data value with index i
+std::vector<float> LQBinarizer::value_bits_as_floats(uint32_t i)
+{
+  std::vector<float> bits(encode_bits);
+
+  int32_t idx = i >> 5;    // divide to 32
+  int32_t offset = i & 31; // mod of 32
+
+  for (int32_t b = 0; b < encode_bits; ++b)
+  {
+    int32_t bit_value = (data_binary[b * data_bin_size + idx] >> offset) & 1;
+    bits[b] = (bit_value == 1) ? 1.0f : -1.0f;
+  }
+
+  return bits;
+}
+
+void LQBinarizer::gradient_descent_scales(const float *data_vector)
```

Comment:
Gradient descent in binarizer looks some strange. Better to move train functionality to `QEM` class. Latter.